src/storage/mod.rs

```diff
@@ -191,6 +191,96 @@ where
     }
 }
 
+/// An entry to a storage which has a component associated to the entity.
+pub struct OccupiedEntry<'a, 'b: 'a, T: 'a, D: 'a> {
+    entity: Entity,
+    storage: &'a mut Storage<'b, T, D>,
+}
+
+impl<'a, 'b, T, D> OccupiedEntry<'a, 'b, T, D>
+    where T: Component,
+          D: Deref<Target = MaskedStorage<T>>,
+{
+    /// Get a reference to the component associated with the entity.
+    pub fn get(&self) -> &T {
+        unsafe { self.storage.data.inner.get(self.entity.id()) }
+    }
+}
+
+impl<'a, 'b, T, D> OccupiedEntry<'a, 'b, T, D>
+    where T: Component,
+          D: DerefMut<Target = MaskedStorage<T>>,
+{
+    /// Get a mutable reference to the component associated with the entity.
+    pub fn get_mut(self) -> &'a mut T {
+        unsafe { self.storage.data.inner.get_mut(self.entity.id()) }
+    }
+
+    /// Inserts a value into the storage and returns the old one.
+    pub fn insert(&mut self, mut component: T) -> T {
+        std::mem::swap(&mut component, unsafe { self.storage.data.inner.get_mut(self.entity.id()) });
+        component
+    }
+
+    /// Removes the component from the storage and returns it.
+    pub fn remove(self) -> T {
+        self.storage.data.remove(self.entity.id()).unwrap()
+    }
+}
+
+/// An entry to a storage which does not have a component associated to the entity.
+pub struct VacantEntry<'a, 'b: 'a, T: 'a, D: 'a> {
+    entity: Entity,
+    storage: &'a mut Storage<'b, T, D>,
+}
+
+impl<'a, 'b, T, D> VacantEntry<'a, 'b, T, D>
+    where T: Component,
+          D: DerefMut<Target = MaskedStorage<T>>,
+{
+
+    /// Inserts a value into the storage.
+    pub fn insert(self, component: T) -> &'a mut T {
+        let id = self.entity.id();
+        self.storage.data.mask.add(id);
+        unsafe {
+            self.storage.data.inner.insert(id, component);
+            self.storage.data.inner.get_mut(id)
+        }
+    }
+}
+
+/// Entry to a storage for convenient filling of components or removal based on whether
+/// the entity has a component.
+pub enum StorageEntry<'a, 'b: 'a, T: 'a, D: 'a> {
+    /// Entry variant that is returned if the entity does has a component.
+    Occupied(OccupiedEntry<'a, 'b, T, D>),
+    /// Entry variant that is returned if the entity does not have a component.
+    Vacant(VacantEntry<'a, 'b, T, D>),
+}
+
+impl<'a, 'b, T, D> StorageEntry<'a, 'b, T, D>
+    where T: Component,
+          D: DerefMut<Target = MaskedStorage<T>>
+{
+
+    /// Inserts a component if the entity does not contain a component.
+    pub fn or_insert(self, component: T) -> &'a mut T {
+        self.or_insert_with(|| component)
+    }
+
+    /// Inserts a component using a lazily called function that is only called
+    /// when inserting the component.
+    pub fn or_insert_with<F>(self, component: F) -> &'a mut T
```

Comment:
I think the `F`-typed argument should be named something other than `component`? stdlib uses `default`